File: java/1160. Find Words That Can Be Formed by Characters/Solution.java
class Solution {
    public int countCharacters(String[] words, String chars) {

        Map<Character, Integer> freq = new HashMap<>();
        int totalLength = 0;

        for (int i = 0; i < chars.length(); i++) {
            freq.put(chars.charAt(i), freq.getOrDefault(chars.charAt(i), 0) + 1);
        }

        for (String word : words) {
            totalLength += helper(word, freq);
        }

        return totalLength;
    }

    public int helper(String word, Map<Character, Integer> freq) {
        Map<Character, Integer> chars = new HashMap<>();

        for (int i = 0; i < word.length(); i++) {
            chars.put(word.charAt(i), chars.getOrDefault(word.charAt(i), 0) + 1);
        }

        for (Character c : chars.keySet()) {
            if (freq.getOrDefault(c, 0) < chars.get(c)) {
                return 0;
            }
        }

        return word.length();

    }
}

File: python/1160. Find Words That Can Be Formed by Characters/solution.py
class Solution:
    def countCharacters(self, words: List[str], chars: str) -> int:


        def helper(word, freq):
            count = [0] * 26

            for char in word:
                count[ord(char) - ord('a')] += 1

            for i in range(len(count)):
                if count[i] > freq[i]:
                    return 0

            return len(word)

        freq = [0] * 26
        totalLength = 0

        for char in chars:
            freq[ord(char) - ord('a')] += 1

        for word in words:
            totalLength += helper(word, freq)

        return totalLength

